File: sorting_algo/merge_sort.c
// Merge Sort Algorithm
// 합병 정렬 알고리즘

#include <stdio.h>
#define ELEMENTS 10

// merge divided list
void merge(int list[], int left, int mid, int right) {
	int temp_list[ELEMENTS];
	int L = left;
	int R = mid + 1;
	int n = left;
	
	// compare value L and value R
	// we want the smaller than mid_value goes to left side of list
	// and bigger than mid_value goes to right side of list
	while (L <= mid && R <= right) {
		// if L is smaller than R,
		if (list[L] <= list[R]) {
			// it goes left side
			temp_list[n++] = list[L++];
		} else {
			// else, it goes right side
			temp_list[n++] = list[R++];
		}
	}
	// if value L is bigger than mid
	if (L > mid) {
		// start point is R(bigger than mid_value) and loop until right
		for (int i = R; i <= right; i++) {
			temp_list[n++] = list[i];
		}
	} else {
		// else, start point is L(smaller than mid_value) and loop until mid
		for (int i = L; i <= mid; i++) {
			temp_list[n++] = list[i];
		}
	}
	// re-sort list
	for (int i = left; i <= right; i++) {
		list[i] = temp_list[i];
	}
}

// this for divide list in half until left >= right
// after finished dividing, starting merge
void merge_sort(int list[], int left, int right) {
	if (left >= right) {
		return;
	}
	
	int mid = (left + right) / 2;
	merge_sort(list, left, mid);
	merge_sort(list, mid + 1, right);
	merge(list, left, mid, right);
}

int main() {
	int list[ELEMENTS] = {5, 3, 2, 7, 8, 1, 9, 0, 6, 4};
	merge_sort(list, 0, 9);
	for (int i = 0; i < 10; i++) {
		printf("%d ", list[i]);
	}
	return 0;
}

File: sorting_algo/merge_sort.py
# Merge Sort Algorithm

def merge(num_list, left, right):
    temp_list = []
    mid = (left + right) // 2
    L = left
    R = mid + 1

    while L <= mid and R <= right:
        if num_list[L] <= num_list[R]:
            temp_list.append(num_list[L])
            L += 1
        else:
            temp_list.append(num_list[R])
            R += 1

    while L <= mid:
        temp_list.append(num_list[L])
        R += 1

    while R <= right:
        temp_list.append(num_list[R])
        R += 1

    for i in range(left, right + 1):
        num_list[i] = temp_list[i - left]

    return num_list


def merge_sort(num_list, left, right):
    if left >= right:
        return

    mid = (left + right) // 2
    merge_sort(num_list, left, mid)
    merge_sort(num_list, mid + 1, right)
    return merge(num_list, left, right)

num_list = [5, 3, 2, 7, 8, 1, 9, 0, 6, 4]
merge_sort(num_list, 0, 9)
print(num_list)
